Improve GC discovery sync performance

ironcladlou opened the pull request:
Improve GC discovery sync performance by only syncing when discovered
resource diffs are detected. Before, the GC worker pool was shut down
and monitors resynced unconditionally every sync period, leading to
significant processing delays causing test flakes where otherwise
reasonable GC timeouts were being exceeded.

Related to https://github.com/kubernetes/kubernetes/issues/49966.

/cc @kubernetes/sig-api-machinery-bugs

```release-note
NONE
```

ironcladlou committed:
Improve GC discovery sync performance

Improve GC discovery sync performance by only syncing when discovered
resource diffs are detected. Before, the GC worker pool was shut down
and monitors resynced unconditionally every sync period, leading to
significant processing delays causing test flakes where otherwise
reasonable GC timeouts were being exceeded.

Related to https://github.com/kubernetes/kubernetes/issues/49966.

ironcladlou commented:
/retest

caesarxuchao commented on pkg/controller/garbagecollector/garbagecollector.go:
```diff
@@ -162,25 +163,59 @@ type resettableRESTMapper interface {
+		}
+
+		// Something has changed, so track the new state and perform a sync.
+		glog.V(5).Infof("syncing garbage collector with updated resources from discovery: %v", newResources)
```
It's worth a `V(2)`

ironcladlou replied:
Done, went ahead and squashed the change in.

caesarxuchao commented on pkg/controller/garbagecollector/garbagecollector.go:
```diff
@@ -162,25 +163,59 @@ type resettableRESTMapper interface {
+
+		// Perform the monitor resync and wait for controllers to report cache sync.
+		//
+		// NOTE: It's possible that newResources will diverge from the resources
```
To clarify, this is not a new issue. The original code has the same inconsistency because the memcached discovery client always get the fresh server resource when `ServerPreferredResources` is called.

caesarxuchao reviewed (commented):
A nit. Otherwise lgtm.

caesarxuchao commented:
[controller manager log](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/50335/pull-kubernetes-e2e-gce-etcd3/45892/artifacts/e2e-45892-master/kube-controller-manager.log) of the e2e test shows that GC is only paused twice.

ironcladlou commented:
GC flakes in the GC flake reduction PR... so meta. I'll investigate.

ironcladlou commented:
"Garbage collector should delete jobs and pods created by cronjob" is really difficult to debug because the logging in that test is using `ginkgo.By` which totally obfuscates the timestamps of the logging statements. I also really need v=5 logs from the GC and graph builder. However, I was able to observe the following sequence.

GC handles job deletion:

```
I0809 13:51:06.792802       5 garbagecollector.go:482] delete object [batch/v1/Job, namespace: e2e-tests-gc-57sdk, name: simple-1502286660, uid: ca694426-7d09-11e7-a500-42010a800002] with propagation policy Background
```

Job observed as being deleted 17s later:

```
I0809 13:51:06.835434       5 wrap.go:42] DELETE /apis/batch/v1/namespaces/e2e-tests-gc-57sdk/jobs/simple-1502286660: (35.738938ms) 200 [[kube-controller-manager/v1.8.0 (linux/amd64) kubernetes/6be86fb/system:serviceaccount:kube-system:generic-garbage-collector] [::1]:41856]
I0809 13:51:23.325290       5 wrap.go:42] GET /apis/batch/v1/namespaces/e2e-tests-gc-57sdk/jobs/simple-1502286660: (221.310638ms) 404 [[kube-controller-manager/v1.8.0 (linux/amd64) kubernetes/6be86fb/system:serviceaccount:kube-system:generic-garbage-collector] [::1]:41856]
```

GC deletes pod, coincides with observed deletion of job:

```
I0809 13:51:23.779751       5 garbagecollector.go:378] processing item [v1/Pod, namespace: e2e-tests-gc-57sdk, name: simple-1502286660-tpfrr, uid: ca8438ba-7d09-11e7-a500-42010a800002]
I0809 13:51:24.815337       5 garbagecollector.go:482] delete object [v1/Pod, namespace: e2e-tests-gc-57sdk, name: simple-1502286660-tpfrr, uid: ca8438ba-7d09-11e7-a500-42010a800002] with propagation policy Background
```

Pod deleted immediately but observed as being deleted 63s later:

```
I0809 13:51:24.875931       5 wrap.go:42] DELETE /api/v1/namespaces/e2e-tests-gc-57sdk/pods/simple-1502286660-tpfrr: (55.930048ms) 200 [[kube-controller-manager/v1.8.0 (linux/amd64) kubernetes/6be86fb/system:serviceaccount:kube-system:generic-garbage-collector] [::1]:41856]
I0809 13:52:18.949854       5 wrap.go:42] GET /api/v1/namespaces/e2e-tests-gc-57sdk/pods/simple-1502286660-tpfrr?resourceVersion=0: (631.667µs) 200 [[kube-apiserver/v1.8.0 (linux/amd64) kubernetes/6be86fb] 127.0.0.1:48910]
I0809 13:52:19.004764       5 wrap.go:42] DELETE /api/v1/namespaces/e2e-tests-gc-57sdk/pods/simple-1502286660-tpfrr: (79.579132ms) 200 [[kubelet/v1.8.0 (linux/amd64) kubernetes/6be86fb] 35.193.32.29:57630]
I0809 13:52:27.311421       5 wrap.go:42] GET /api/v1/namespaces/e2e-tests-gc-57sdk/pods/simple-1502286660-tpfrr: (4.435174ms) 404 [[kubelet/v1.8.0 (linux/amd64) kubernetes/6be86fb] 35.193.32.29:57630]
```

I don't understand the second `DELETE`, but it looks like the pod doesn't start 404'ing until possibly 13:52:27, 63s later. I'm not sure why there's such a long delay between the DELETEs and the API server 404'ing GETs of those resources.

ironcladlou commented:
I'm going to spend a bit of time trying to reproduce one of these flakes locally. Might get lucky.

caesarxuchao commented:
> I'm not sure why there's such a long delay between the DELETEs and the API server 404'ing GETs of those resources.

It looks like the pod is has a long graceful deletion period. The job podTemplate needs to set the [TerminationGracefulPeriod](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/types.go#L2441) to 0

ironcladlou committed:
Use zero TerminationGracePeriodSeconds in fixture

Pods associated with the test JobTemplate should use a zero
TerminationGracePeriodSeconds to ensure they're deleted immediately.
This should improve test timing assumption consistency.

ironcladlou commented:
>It looks like the pod is has a long graceful deletion period. The job podTemplate needs to set the TerminationGracefulPeriod to 0

Fixed, thanks.

The "Garbage collector should not delete dependents that have both valid owner and owner that's waiting for dependents to be deleted" might be https://github.com/kubernetes/kubernetes/issues/42086.

caesarxuchao commented:
/lgtm

ironcladlou commented:
[Two more examples](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/50335/pull-kubernetes-e2e-gce-etcd3/46086/build-log.txt) of https://github.com/kubernetes/kubernetes/issues/42086 in [this test run](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/50335/pull-kubernetes-e2e-gce-etcd3/46086/).

```
I0809 18:19:21.111] [It] should keep the rc around until all its pods are deleted if the deleteOptions says so
I0809 18:19:21.111]   /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/apimachinery/garbage_collector.go:673
I0809 18:19:21.111] [1mSTEP[0m: create the rc
I0809 18:19:21.111] Aug  9 18:18:37.584: INFO: failed to wait for the rc.Status.Replicas to reach rc.Spec.Replicas: timed out waiting for the condition
```

```
I0809 18:19:21.111] [It] should keep the rc around until all its pods are deleted if the deleteOptions says so
I0809 18:19:21.111]   /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/apimachinery/garbage_collector.go:673
I0809 18:19:21.111] [1mSTEP[0m: create the rc
I0809 18:19:21.111] Aug  9 18:18:37.584: INFO: failed to wait for the rc.Status.Replicas to reach rc.Spec.Replicas: timed out waiting for the condition
```

ironcladlou commented:
/retest

caesarxuchao commented:
All the modifications to `test/` are for gc tests. The `cmd/kube-controller-manager` change is an one-line change that sets up the gc. I'll manually apply the approval label.
